### DISCUSSION TITLE
cumulated reward

### DISCUSSION THREAD

**Discussion by akangw** [ds4dm/ml4co-competition#47] [General]
Hi folks,

I have a naive question. In the primal track, based on the primal integral metric https://www.ecole.ai/2021/ml4co-competition/, we want to minimize the primal integral. However, when I ran my code and found "cumulated reward (to be maximized): XXXX". Does anybody have a clue about why it is 'maximize'? Can @gasse provide some help? Thanks in advance!

Best,
Vincent.

**Comment by gasse** [comment]
Hi @akangw ,

Thank you for asking, naive questions are often very good questions. I realize this can be confusing, let me try to clear things up and give you the rationale behind it.

The primal integral metric is indeed to be minimized. However, in the code we tried to formulate all three control problems (primal, dual and config) within a reinforcement learning (RL) framework, where maximization is more standard. RL traditionally relies on the concept of **rewards** received at each time step, so that the goal is to maximize the **cumulated reward**. Therefore, instead of minimizing the primal integral, we now maximize the negated primal integral. The change of sign can be see in the code here:
https://github.com/ds4dm/ml4co-competition/blob/973315963e0e04925275ed1dc13dd49360597a0d/common/evaluate.py#L102

We had to choose between either:
1. directly reporting the metrics as described in the competition's website. This would have turned rewards to be maximized into penalties to be minimized, which does not stick to the usual RL formulation.
2. reporting the negated version of the metrics, to stick to a standard RL formulation. This adds maybe a little bit of confusion, but is more friendly to people used to RL as they can readily plug those rewards into standard RL algorithms for training, without giving much thought to it.

As you noticed, we went for option 2.

I hope this was helpful. Don't hesitate to ask us other questions if you have any :)

Best,
Maxime

**Reply by akangw** [reply]
Oh, @gasse thank you so much for your help. That makes sense now. 

Best,
Vincent.

**Reply by akangw** [reply]
@gasse I have a follow-up question. In the primal track, when my policy returns an action, how do I know whether my action leads to a feasible solution (it does not necessarily improve the primal bound, I just want to know whether it is feasible but I do not want to solve an LP by myself)? Do you have any idea? Thanks in advance!

Best,
Vincent.

**Reply by gasse** [reply]
Hi @akangw , I believe you can use the [following methods](https://github.com/scipopt/PySCIPOpt/blob/master/src/pyscipopt/scip.pyx#L4196-L4231) from PySCIPOpt:
```python
m = model.as_pyscipopt()
m.getNSols()
m.getNSolsFound()
m.getNLimSolsFound()
m.getNBestSolsFound()
m.getSols()
m.getBestSol()
```
Best,
Maxime

**Reply by akangw** [reply]
Hi @gasse ,

Thank you! I have tried this before and noticed that if my returned action is feasible but has a worse objective than the incumbent, then it will not be counted as a solution. As a result, these functions will not tell me whether a returned action is feasible or not.

Best,
Vincent.

**Reply by gasse** [reply]
I looked a bit into it, and unfortunately it seems impossible to extract that information from the environment. Indeed, the primal environment relies on [`SCIPtrySol`](https://www.scipopt.org/doc/html/scip__sol_8c_source.php#l03126), which will not even check the feasibility of the candidate solution if it can already concludes that the solution's objective value is not high enough to be of interest (see [here](https://www.scipopt.org/doc/html/primal_8c_source.php#l01518)). In that case, the complete solution recovered after solving the LP is discarded.

A workaround would be to allow you (the user) to solve the LP on your side and provide not a partial but a complete solution to the environment, so that the environment does not have to solve the LP a second time but simply try the solution. I think this could be doable without introducing any breaking change to the competition's API. Let me discuss with the other organizers whether we can include that feature or not in the environment.

Best,
Maxime

**Reply by akangw** [reply]
Hi @gasse ,

I really appreciate your efforts on this. Yes, I also did not find a better way out. I think, in a RL framework, 'feasibility' of an action is also part of 'reward' and hence should be meaningful information to know.

Best,
Vincent.

**Reply by gasse** [reply]
I just checked the [Ecole code](https://github.com/ds4dm/ecole/blob/6cc08bb2a8dfff1d15b3e907e67a6ac22f6d9ee9/libecole/src/dynamics/primalsearch.cpp#L51). I thought you could only provide partial solutions with values for integer variables, but it turns out we only check that variable ids are in range, not that they correspond to integer variables necessarily. This is good news for you, if you manage to solve an LP (and check for feasibility) on your side you'll get a complete solution, and the environment's LP will be instantaneous to solve since all variable will be fixed. You should be able to do so by making a copy of the SCIP model in your first interaction with the environment, and then fix variables and solve LPs in probing mode. You'll find ressources to help you do that [here](https://github.com/ds4dm/ml4co-competition/discussions/31#discussioncomment-1043818) (how model copy works) and [here](https://github.com/ds4dm/ecole/blob/master/libecole/src/dynamics/primalsearch.cpp#L68-L99) (how probing mode works).

Best,
Maxime

**Reply by akangw** [reply]
Hi @gasse ,

that is interesting. Thanks for the info. 

Best,
Vincent.
